bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthStoreHandlerImpl.java

```diff
@@ -0,0 +1,531 @@
+ *
+ * @author Gary Tse - Initial Contribution
+ *
+ */
+@Component(service = OAuthStoreHandler.class, property = {
+        "CIPHER_TARGET=org.eclipse.smarthome.auth.oauth2client.internal.cipher.SymmetricKeyCipher" })
+public class OAuthStoreHandlerImpl implements OAuthStoreHandler {
+
+    // easy mocking with protected access
+    protected static final int EXPIRE_DAYS = 183;
+    protected static final int ACCESS_TOKEN_CACHE_SIZE = 50;
+
+    private static final String STORE_NAME = "StorageHandler.For.OAuthClientService";
+
+    @NonNull
+    private static final String STORE_KEY_INDEX_OF_HANDLES = "INDEX_HANDLES";
+
+    private static final String EXCEPTION_STORAGE = "Unable to continue, storage is not available";
+
+    private StorageService _storageServiceDoNotUseDirectly;
+    private Storage<String> _storeDoNotUseDirectly;
+
+    private HashSet<String> allHandles = new HashSet<>(); // must be initialized
+
+    // gets replaced if Store becomes available. storageFacade handles all the times when
+    // the actual storage becomes unavilable and locking issues.
+    private StorageFacade storageFacade = new StorageFacade(null);
+
+    private Cache<String, AccessTokenResponse> accessTokenCache;
+
+    private final LinkedHashSet<StorageCipher> allAvailableStorageCiphers = new LinkedHashSet<>();
+    private Optional<StorageCipher> storageCipher = Optional.ofNullable(null);
+
+    private final Logger logger = LoggerFactory.getLogger(OAuthStoreHandlerImpl.class);
+
+    public OAuthStoreHandlerImpl() {
+        // The cache needs to be initialized at constructor.
+        // This is because the setStorageService uses the accessTokenCache
+        //
+        // 50 cached access token entries, remove after half year if not used
+        accessTokenCache = CacheBuilder.newBuilder().maximumSize(ACCESS_TOKEN_CACHE_SIZE)
+                .expireAfterAccess(EXPIRE_DAYS, TimeUnit.DAYS).removalListener(new StorageEntryRemovalListener())
+                .build();
+    }
+
+    @Activate
+    public void activate(Map<String, Object> properties) {
+        String cipherTarget = (String) properties.getOrDefault("CIPHER_TARGET", SymmetricKeyCipher.CIPHER_ID);
+
+        // choose the cipher by the cipherTarget
+        storageCipher = allAvailableStorageCiphers.stream()
+                .filter(cipher -> cipher.getUniqueCipherId().equals(cipherTarget)).findFirst();
+
+        logger.debug("Using Cipher: {}", storageCipher.orElse(null));
+        // this allows future implementations to change cipher by just setting the CIPHER_TARGET
+
+        // by now the storage and cipher should be ready, load all tokens
+        loadAllHandlesAndTokens();
+    }
+
+    @Deactivate
+    public void deactivate() {
+        accessTokenCache.cleanUp(); // this removes old entries
+        accessTokenCache = null;
+        storageFacade.close(); // this removes old entries
+        // storageFacade = null; // do not set to null
```

Comment:
I believe this is redundant. If you want to specify the reason behind not setting to null, you can add a comment to the javadoc section of this method

Reply:
Added to comments.

Reply:
If you have already added to the comments, you can remove the comments here (line nos 144-145).